### ISSUE TITLE
[GPU] The topology fingerprint is process-local: it hashes one local device's interconnect info, so the processes of one job fingerprint the same topology differently

### ISSUE THREAD

**Issue by AshishKumar4** [openxla/xla#49371]
### Description

`StreamExecutorGpuTopologyDescription::Fingerprint()` (`xla/pjrt/gpu/se_gpu_topology_description.cc`)
hashes the deterministic serialization of the client's `GpuTopologyProto`. That proto embeds
`gpu_target_config.gpu_device_info` for one local device, including `device_interconnect_info`
(`GpuDeviceInfoProto` field 29: `active_links`, `cluster_uuid`, `clique_id`). `cuda_executor.cc`
reads it from NVML for each device. So the fingerprint isn't a property of the job's topology: it
depends on which GPU is the process's first device.

On one host with 4x RTX 3090 (GPU0-GPU1 NV4, GPU2-GPU3 PHB, cross-pair SYS), a job of four
processes, one GPU each, builds the same global topology in every process: 4 devices, 4 processes.
Yet processes on GPU0 and GPU1 get one fingerprint, and processes on GPU2 and GPU3 get another.
Decoding each process's topology proto shows that one field differs:
`gpu_target_config.gpu_device_info.device_interconnect_info.active_links` is 4 on GPU0 and GPU1 and
absent on GPU2 and GPU3 (`topology-diff.txt` below). `Serialize()` isn't canonical either: map
entries come out in a different order in every process, although `Fingerprint()` serializes
deterministically.

### Why it matters

JAX keys its persistent compilation cache by this fingerprint (`jax/_src/cache_key.py`,
`_hash_accelerator_config`, since #39387). XLA's compile of a program that spans processes shards autotuning over
them (`ConfigAssignerPass` with the compile's `MultiProcessKeyValueStore`), and each process waits
for the others' shares. With keys that differ between the processes of one job, some processes load
a cached executable while the others compile, and those compiles wait forever for shares that never
come: jax-ml/jax#40940. JAX can avoid this by hashing the fingerprints of all the processes a
computation spans, as proposed there (branch `cache-key-pool-fingerprints` on AshishKumar4/jax).
Anything else that uses `Fingerprint()` as a job-wide identity has the same problem.

Even with that fix, a program on one process's devices keeps that process's fingerprint. In a
four-process training job on this host, the processes on GPU2 and GPU3 recompile their one-process
programs on every run, because only process 0 writes the cache and their fingerprint differs from
its. A one-device program reads no link at all.

### Questions

1. Should the topology fingerprint of a multi-process client describe the job rather than one local
   device? The client already exchanges local topologies when it's created (the key-value exchange
   that builds the global topology). A fingerprint over every device's description would be the same
   in every process.
2. Should per-device link info be in a key for a program that spans processes? At openxla/xla@91888df6
   (jax 0.11.2), `active_links` reaches compilation in three places:
   - `SolLatencyEstimator`, used only on Hopper, Blackwell and MI300 (`IsSupportedForModule`);
   - the analytical latency estimator (`gpu_collective_performance_model.cc`), behind
     `xla_gpu_enable_analytical_latency_estimator`, which is off by default;
   - the collective kernels (`all_reduce.cc`, `all_gather.cc`), which need NVLink and replicas all
     local to one process, which a multi-process program never has.

   #40177 made `GpuTargetConfig` device-agnostic for exactly this reason: fields unique to one GPU,
   the PCI bus ID and the NVLink UUIDs, made the configs of two GPUs of one model differ, "which is
   a problem when used as a cache key". It left `active_links`, which is now the one field that
   differs here. #49099 (`openxla/xla@86ee39e615`) normalizes `device_memory_size` and clears
   `model_str` for the same portability. `active_links` can't simply be dropped, since a
   single-process program's collective kernels and SOL estimates read it. But a program that spans
   processes doesn't read it at this revision, so a fingerprint that describes the job, or a link
   count taken as the job's minimum, would keep one key per job.
3. The target config describes only one device of a process. A single process that owns GPUs with
   different links (all four GPUs of this host) compiles every program with its first device's
   `active_links`, including a program that runs on GPU2 and GPU3 alone. Is that intended?

### Evidence

<details><summary>topology-diff.txt: each rank's decoded topology proto and the one differing field</summary>

```
Raw-protobuf decode of xla_client.get_topology_for_devices(jax.devices()).serialize() in each process of a
four-process pool (dew launch 4x1) on the 4x RTX 3090 box (GPU0-GPU1 NV4, GPU2-GPU3 PHB, cross-socket SYS), jax 0.11.2:
  rank 0 (GPU0): 766 bytes, 80 fields    rank 1 (GPU1): 766 bytes, 80 fields
  rank 2 (GPU2): 761 bytes, 79 fields    rank 3 (GPU3): 761 bytes, 79 fields
As field sets, ranks 0 and 1 are identical and ranks 2 and 3 are identical. The one difference:
  field path /9/2/8/1/29/1 = 4 on ranks 0 and 1, absent on ranks 2 and 3
  (inside the embedded GpuTopologyProto's device description; 4 matches the NV4 link count between GPU0 and GPU1).
The serialized bytes also list a repeated/map field's entries in a different order in every process
(rank 0 and rank 1 differ in order only), so serialize() is not canonical either.
JAX's cache key hashes topology.fingerprint() (jax/_src/cache_key.py _hash_accelerator_config); in the same pool
its component hash was d8852249... on ranks 0,1 and 8b8b9014... on ranks 2,3 for one 4-device SPMD jit
(computation, compile_options, XLA flags and backend version hashes identical on all ranks).
In a single process the four devices give one accelerator_config hash (dc2d2ee7...).
```
</details>

<details><summary>JAX's cache-key components for one four-process SPMD program, first 12 hex digits of each component's hash (DEBUG log of jax._src.cache_key, jax 0.11.2)</summary>

```
component          rank 0 (GPU0)  rank 1 (GPU1)  rank 2 (GPU2)  rank 3 (GPU3)
computation        88c32b69db78   88c32b69db78   88c32b69db78   88c32b69db78 
jax_lib version    7558e7d2c821   7558e7d2c821   7558e7d2c821   7558e7d2c821 
backend version    a224da3101c0   a224da3101c0   a224da3101c0   a224da3101c0 
XLA flags          e3b0c44298fc   e3b0c44298fc   e3b0c44298fc   e3b0c44298fc 
compile_options    b47e2d3afdb0   b47e2d3afdb0   b47e2d3afdb0   b47e2d3afdb0 
accelerator_config d8852249e57e   d8852249e57e   8b8b901463e0   8b8b901463e0 
compression        0ea55c28f801   0ea55c28f801   0ea55c28f801   0ea55c28f801 
custom_hook        e3b0c44298fc   e3b0c44298fc   e3b0c44298fc   e3b0c44298fc 
```

Only the accelerator config differs, between ranks 0,1 and ranks 2,3.
</details>

### System info

- jax 0.11.2 and jaxlib 0.11.2 (jax-cuda12-plugin), XLA 91888df6ce85102c30220e41d952065925e10886
- 4x NVIDIA GeForce RTX 3090, driver 595.84, one host. `nvidia-smi topo -m`: GPU0-GPU1 NV4,
  GPU2-GPU3 PHB, cross-pair SYS
